core/services/run_queue.go

```diff
@@ -34,81 +36,188 @@ type RunQueue interface {
-				delete(rq.workers, runID)
-				numberRunQueueWorkers.Set(float64(len(rq.workers)))
-				rq.workersMutex.Unlock()
-				break
+		case jobSpecID := <-rq.chWorkerDone:
+			delete(rq.workers, jobSpecID)
+			if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+				panic("numWorkers should never be < 0")
 			}
-			rq.workers[runID] = queueCount - 1
-			rq.workersMutex.Unlock()
 
-			if err := rq.runExecutor.Execute(run.ID); err != nil {
-				logger.Errorw(fmt.Sprint("Error executing run ", runID), "error", err)
+		case <-rq.chStop:
+			for _, worker := range rq.workers {
+				worker.chStop <- struct{}{}
+				<-worker.chDone
+				if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+					panic("numWorkers should never be < 0")
+				}
 			}
+			return
 		}
+	}
+}
 
-		rq.workersWg.Done()
+func (rq *runQueue) runSingleJobSpecWorker(jobSpecID models.ID, worker *singleJobSpecWorker) {
+	defer close(worker.chDone)
+
+	promNumberRunQueueWorkersStarted.Inc()
+
+	var (
+		startOnce       sync.Once
+		chWorkerStarted = make(chan struct{})
+		chResume        = make(chan []models.ID)
+		chWorkComplete  = make(chan struct{})
+	)
+
+	// The worker goroutine accepts job run requests from the coordinator loop below.  It
+	// can accept further requests after work has begun.  Once its queue of work is empty,
+	// it sends a "work finished" message to that loop, which may or may not cause it to
+	// spin down, depending on whether further requests are already enqueued.
+	go func() {
+		for runsRequested := range chResume {
+			// We have to wait until the worker has actually received a job before indicating
+			// to the coordinator loop that work has started.  If we do this before the worker's
+			// for loop starts, there's a race that can allow the coordinator to receive a chStop
+			// message and shut down the worker even though runs are queued.
+			startOnce.Do(func() { close(chWorkerStarted) })
+
+			for _, runID := range runsRequested {
+				select {
+				case <-worker.chStop:
+					return
+				default:
+				}
+
+				promNumberRunsQueued.Inc()
+
+				if err := rq.runExecutor.Execute(&runID); err != nil {
+					logger.Errorw(fmt.Sprint("Error executing run ", runID.String()), "error", err)
+				}
+			}
+			select {
+			case chWorkComplete <- struct{}{}:
+			case <-worker.chStop:
+				return
+			}
+		}
 	}()
+
+	// The coordinator loop accepts "do job run" requests from orchestrateWorkers, "work finished"
+	// signals from the worker goroutine, and "stop" messages from the Chainlink application,
+	// coordinating them such that we can avoid races when messages arrive simultaneously.
+	var runsRequested []models.ID
+	for {
+		select {
+		case run := <-worker.chAddRun:
+			if !bytes.Equal(run.JobSpecID.Bytes(), jobSpecID.Bytes()) {
+				panic("worker received a run request from another job spec")
+			}
+
+			runsRequested = append(runsRequested, *run.ID)
+
+			select {
+			case chResume <- runsRequested:
+				runsRequested = nil
+			case <-chWorkerStarted:
```

Comment:
I don't understand how this achieves the behavior described below. Is there a test that covers this? It seems like it is never used, only ever closed... I don't understand the situation where it does anything.

Reply:
Closing a `chan struct{}` is a common Go pattern that allows you to broadcast a single, simple message (usually to indicate startup or shutdown).  In this particular case, there's no difference between `chWorkerStarted <- struct{}{}` and `close(chWorkerStarted)`.  I chose `close` because 1) it's more common in Go codebases for cases like these, and 2) it indicates that the signal can only be sent once.

Re: how this achieves the desired behavior: as soon as the channel is closed, `<-chWorkerStarted` unblocks, which is how we avoid hitting the `chStop` case before enqueued work can be distributed.

Re: tests, yes -- the tests actually forced me to implement this, because they enqueue work and then call `.Stop()` so quickly.  Without this check, no work gets done because the `stop` message sometimes happens too soon (non-deterministically).